axum-extra/src/response/file_stream.rs

```diff
@@ -0,0 +1,270 @@
+        let file = File::open(&path).await?;
+        let mut content_size = None;
+        let mut file_name = None;
+
+        // get file metadata length
+        if let Ok(metadata) = file.metadata().await {
+            content_size = Some(metadata.len());
+        }
+
+        // get file name
+        if let Some(file_name_os) = path.file_name() {
+            if let Some(file_name_str) = file_name_os.to_str() {
+                file_name = Some(file_name_str.to_owned());
+            }
+        }
+
+        // return FileStream
+        Ok(FileStream {
+            stream: ReaderStream::new(file),
+            file_name,
+            content_size,
+        })
+    }
+
+    /// Set the file name of the file.
+    pub fn file_name<T: Into<String>>(mut self, file_name: T) -> Self {
+        self.file_name = Some(file_name.into());
+        self
+    }
+
+    /// Set the size of the file.
+    pub fn content_size<T: Into<u64>>(mut self, len: T) -> Self {
+        self.content_size = Some(len.into());
+        self
+    }
+}
+
+impl<S> IntoResponse for FileStream<S>
+where
+    S: TryStream + Send + 'static,
+    S::Ok: Into<Bytes>,
+    S::Error: Into<BoxError>,
+{
+    fn into_response(self) -> Response {
+        let mut resp = Response::builder().header(header::CONTENT_TYPE, "application/octet-stream");
+
+        if let Some(file_name) = self.file_name {
+            resp = resp.header(
+                header::CONTENT_DISPOSITION,
+                format!("attachment; filename=\"{}\"", file_name),
+            );
+        };
+
+        if let Some(content_size) = self.content_size {
+            resp = resp.header(header::CONTENT_LENGTH, content_size);
+        };
+
+        resp.body(body::Body::from_stream(self.stream))
+            .unwrap_or_else(|e| {
+                (
+                    StatusCode::INTERNAL_SERVER_ERROR,
+                    format!("build FileStream responsec error:{}", e),
+                )
+                    .into_response()
+            })
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use axum::{extract::Request, routing::get, Router};
+    use body::Body;
+    use http_body_util::BodyExt;
+    use std::io::Cursor;
+    use tokio_util::io::ReaderStream;
+    use tower::ServiceExt;
+
+    #[tokio::test]
+    async fn response_file_stream() -> Result<(), Box<dyn std::error::Error>> {
+        let app = Router::new().route(
+            "/file",
+            get(|| async {
+                // Simulating a file stream
+                let file_content = b"Hello, this is the simulated file content!".to_vec();
+                let size = file_content.len() as u64;
+                let reader = Cursor::new(file_content);
+
+                // response file stream
+                let stream = ReaderStream::new(reader);
+                FileStream::new(stream)
+                    .file_name("test")
+                    .content_size(size)
+                    .into_response()
```

Comment:
Should we add more tests:
- one where the filename is not set
- one where the content size is not set
- one where the filename and the content size are not set

Reply:
> Should we add more tests:
> 
> * one where the filename is not set
> * one where the content size is not set
> * one where the filename and the content size are not set

Yeah, I added more tests.